File: compiler/tokenizer.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum

@dataclass 
class Token:
    class TOKEN_TYPE(Enum):
        IDENTIFIER = 0
        LITERAL = 1
        TYPE = 2
    
        ADD = 3
        SUBTRACT = 4
        MULTIPLY = 5
        DIVIDE = 6
        MODULUS = 7

        EQUALS = 17
        GREATER_THAN = 18
        LESS_THAN = 19
        AND = 20
        OR = 21
    
        IF = 12
        WHILE = 13
        END = 14
        ELIF = 23
        ELSE = 24
        RETURN = 25

        OPENING_PARENTHESIS = 15
        CLOSING_PARENTHESIS = 16
        SEPERATOR = 22

    token_type: TOKEN_TYPE
    value: str

def tokenize(lines: list[str]) -> list[list[Token]]:
    tokenized_lines: list[list[Token]] = []
    in_quotes = False
    word_collecter = ""

    for line in lines:
        tokenized_lines.append([])

        def add_word_token(word: str):
            token_type = Token.TOKEN_TYPE.IDENTIFIER

            if word.replace(".", "").isnumeric():
                token_type = Token.TOKEN_TYPE.LITERAL

            if word == "false" or word == "true":
                token_type = Token.TOKEN_TYPE.LITERAL

            elif word == "String":
                token_type = Token.TOKEN_TYPE.TYPE

            elif word == "Integer":
                token_type = Token.TOKEN_TYPE.TYPE

            elif word == "Decimal":
                token_type = Token.TOKEN_TYPE.TYPE

            elif word == "Boolean":
                token_type = Token.TOKEN_TYPE.TYPE

            elif word == "Void":
                token_type = Token.TOKEN_TYPE.TYPE

            elif word == "if":
                token_type = Token.TOKEN_TYPE.IF

            elif word == "while":
                token_type = Token.TOKEN_TYPE.WHILE

            elif word == "end":
                token_type = Token.TOKEN_TYPE.END

            elif word == "elif":
                token_type = Token.TOKEN_TYPE.ELIF

            elif word == "else":
                token_type = Token.TOKEN_TYPE.ELSE

            elif word == "return":
                token_type = Token.TOKEN_TYPE.RETURN

            # If it's still an identiter we want to add a prefix to it
            if token_type == Token.TOKEN_TYPE.IDENTIFIER:
                word = f"esll_{word}"

            tokenized_lines[-1].append(Token(token_type, word))

        for i in range(len(line)):
            c = line[i]

            added_to_word = False
            token = Token(Token.TOKEN_TYPE.IDENTIFIER, "")

            if in_quotes:
                tokenized_lines[-1][-1].value += c

                if c == "\"":
                    in_quotes = False
                    continue

            elif c == " ":
                pass

            elif c == ",":
                token = Token(Token.TOKEN_TYPE.SEPERATOR, c)

            elif c == "+":
                token = Token(Token.TOKEN_TYPE.ADD, c)

            elif c == "-":
                token = Token(Token.TOKEN_TYPE.SUBTRACT, c)

            elif c == "*":
                token = Token(Token.TOKEN_TYPE.MULTIPLY, c)

            elif c == "/":
                token = Token(Token.TOKEN_TYPE.DIVIDE, c)

            elif c == "%":
                token = Token(Token.TOKEN_TYPE.MODULUS, c)

            elif c == "&":
                token = Token(Token.TOKEN_TYPE.AND, c)

            elif c == "|":
                token = Token(Token.TOKEN_TYPE.OR, c)

            elif c == ">":
                token = Token(Token.TOKEN_TYPE.GREATER_THAN, c)

            elif c == "<":
                token = Token(Token.TOKEN_TYPE.LESS_THAN, c)

            elif c == "=":
                token = Token(Token.TOKEN_TYPE.EQUALS, c)

            elif c == "(":
                token = Token(Token.TOKEN_TYPE.OPENING_PARENTHESIS, c)

            elif c == ")":
                token = Token(Token.TOKEN_TYPE.CLOSING_PARENTHESIS, c)

            elif c == "\"":
                in_quotes = True
                token = Token(Token.TOKEN_TYPE.LITERAL, "\"")

            else: 
                added_to_word = True
                word_collecter += c

            if not added_to_word and len(word_collecter) > 0:
                add_word_token(word_collecter)
                word_collecter = ""

            if not (token.token_type == Token.TOKEN_TYPE.IDENTIFIER and token.value == ""):
                tokenized_lines[-1].append(token)
        
        if len(word_collecter) > 0:
            add_word_token(word_collecter)
            word_collecter = ""

    return tokenized_lines

File: src/tokenizer.cpp
#include "tokenizer.hpp"

#include <ctype.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

const int MAX_TOKENS_PER_LINE = 100;

bool isNumeric(const std::string &str) {
  for (char ch : str) {
    if (ch == '.') {
      continue;
    }
    if (!isdigit(ch)) {
      return false; // Return false if any character is not a digit
    }
  }
  return true; // All characters are digits
}

static Token addWordToken(std::string word) {
  TokenType tokenType = TokenType_IDENTIFIER;

  // Check if it"s a number (integer or float)
  if (isNumeric(word)) {
    tokenType = TokenType_LITERAL;
  }
  // Check if it"s a boolean literal
  else if (word == "false" || word == "true") {
    tokenType = TokenType_LITERAL;
  }
  // Check for specific types
  else if (word == "String" || word == "Integer" || word == "Decimal" ||
           word == "Boolean" || word == "Void") {
    tokenType = TokenType_TYPE;
  }
  // Check for keywords
  else if (word == "if") {
    tokenType = TokenType_IF;
  } else if (word == "while") {
    tokenType = TokenType_WHILE;
  } else if (word == "end") {
    tokenType = TokenType_END;
  } else if (word == "elif") {
    tokenType = TokenType_ELIF;
  } else if (word == "else") {
    tokenType = TokenType_ELSE;
  } else if (word == "return") {
    tokenType = TokenType_RETURN;
  }

  // If it"s still an identifier, prefix it
  if (tokenType == TokenType_IDENTIFIER) {
    word = "esll_" + word;
  }

  return {tokenType, word};
}

std::vector<std::vector<Token>> tokenize(std::vector<std::string> lines) {
  std::vector<std::vector<Token>> tokenizedLines;

  bool inQuotes = false;
  std::string wordCollector = "";

  for (int lineNumber = 0; lineNumber < lines.size(); lineNumber++) {
    std::vector<Token> tokenizedLine;
    const std::string &line = lines[lineNumber];
    int endingCharNumber = 0;

    for (int charNumber = 0; charNumber < line.size(); charNumber++) {
      std::string c(1, line[charNumber]);

      bool addedToWord = false;
      Token token = {TokenType_IDENTIFIER, ""};
      endingCharNumber = charNumber;

      if (inQuotes) {
        tokenizedLine.back().value += c;
        if (c == "\"") {
          inQuotes = false;
          continue;
        }
      }

      else if (c == " ") {
        // Whitespace, no nothing
      } else if (c == ",") {
        token = {TokenType_SEPERATOR, c};
      } else if (c == "+") {
        token = {TokenType_ADD, c};
      } else if (c == "-") {
        token = {TokenType_SUBTRACT, c};
      } else if (c == "*") {
        token = {TokenType_MULTIPLY, c};
      } else if (c == "/") {
        token = {TokenType_DIVIDE, c};
      } else if (c == "%") {
        token = {TokenType_MODULUS, c};
      } else if (c == "&") {
        token = {TokenType_AND, c};
      } else if (c == "|") {
        token = {TokenType_OR, c};
      } else if (c == ">") {
        token = {TokenType_GREATER_THAN, c};
      } else if (c == "<") {
        token = {TokenType_LESS_THAN, c};
      } else if (c == "=") {
        token = {TokenType_EQUALS, c};
      } else if (c == "(") {
        token = {TokenType_OPENING_PARENTHESIS, c};
      } else if (c == ")") {
        token = {TokenType_CLOSING_PARENTHESIS, c};
      } else if (c == "\"") {
        inQuotes = true;
        token = {TokenType_LITERAL, "\""};
      } else {
        addedToWord = true;
        wordCollector += c;
      }

      if (!addedToWord && wordCollector.size() > 0) {
        Token newToken = addWordToken(wordCollector);
        wordCollector = "";
        tokenizedLine.push_back(newToken);
      }

      if (!(token.tokenType == TokenType_IDENTIFIER && token.value == "")) {
        tokenizedLine.push_back(token);
      }
    }

    if (wordCollector.size() > 0) {
      Token newToken = addWordToken(wordCollector);
      wordCollector = "";
      tokenizedLine.push_back(newToken);
    }

    tokenizedLines.push_back(tokenizedLine);
  }
  return tokenizedLines;
}
